src/networkapi.jl

```diff
@@ -183,6 +183,135 @@ function netstoichmat(rn; smap=speciesmap(rn))
 end
 
 
+######################## some reacion complexes matrices and reaction rates ###############################
+"""
+reaction_complexes(rn)
+The function returns a matrix of unique complexes
+    that represent the reaction network,
+reaction_complexes(rn)[:,1] returns Vector of
+complexes and corresponding stoichiometric coefficient
+
+reaction_complexes(rn)[:,2] returns a vector of linear combination of complexes(coefficients being corresponding stoichiometric coefficient)
+"""
+function reaction_complexes(rn)
+    complexes_mat = [];         complexes = [];
+    r = reactions(rn);
+    for i in 1:numreactions(rn)
+        push!(complexes_mat, [r[i].substrates r[i].substoich])
+        push!(complexes_mat, [r[i].products r[i].prodstoich])
+        if r[i].substrates == Any[]
+            push!(complexes, 0)
+            push!(complexes, sum(r[i].products.*r[i].prodstoich))
+        elseif r[i].products == Any[]
+            push!(complexes, sum(r[i].substrates.*r[i].substoich))
+            push!(complexes, 0)
+        else
+            push!(complexes, sum(r[i].substrates.*r[i].substoich))
+            push!(complexes, sum(r[i].products.*r[i].prodstoich))
+        end
+    end
+    return [complexes_mat[unique(i -> complexes[i] , eachindex(complexes))] unique(complexes)]
+end
+
+
+"""
+reaction_rates(rn)
+The function returns a vector of rate of the
+    reactions
+"""
+function reaction_rates(rn)
+    kd = [];
+    r  = reactions(rn)
+    for i in 1:numreactions(rn)
+        push!(kd, r[i].rate)
+    end
+    return kd
+end
+
+
+"""
+The function complex_stoich_matrix(rn) returns a
+    complex stoichiometric matrix of size
+    numspecies x num_complexes
+
+    The k'th column of the matrix denotes the composition of the k-th complex
+"""
+function complex_stoich_matrix(rn; cmp_mat = reaction_complexes(rn)[:,1],
```

Comment:
Similarly, if `reaction_complexes` already switches to representing species by integers, then you don't actually need to use `speciesmap` here right?

Reply:
small demo, based on https://github.com/SciML/Catalyst.jl/pull/359#discussion_r680366630 , for `complex_stoich_matrix`
```julia
function complex_stoich_matrix(rn; cmp_mat = reaction_complexes(rn))
    Z = zeros(Int64, numspecies(rn), length(cmp_mat));
    for i in 1:length(cmp_mat) # traversing along the columns
        if cmp_mat[i].first != Int64[] # NOT A NULL COMPLEX
            Z[cmp_mat[i].first,i] .= cmp_mat[i].second
        end
    end
    return Z
end

```